### PR TITLE
Add singularity support.

### PR DESCRIPTION
**Pull request by olifre**
This also adds support for MOUNT_UNDER_SCRATCH,
which is needed for containers in general.

### CODE REVIEW

File: manifests/config/worker.pp [hunk 1]
```diff
@@ -32,6 +41,17 @@
     notify  => Exec['/usr/sbin/condor_reconfig'],
   }
 
+  if $use_singularity {
+    file { '/etc/condor/config.d/50_singularity.config':
+      content => template($template_singularity),
+      require => Package['condor'],
+      owner   => $condor_user,
```
Comment (kreczko) [inline manifests/config/worker.pp]: ERROR: trailing whitespace found on line 48

File: manifests/params.pp [hunk 2]
```diff
@@ -72,6 +72,7 @@
 
   $pool_home                      = hiera('pool_home', '/pool')
   $pool_create                    = hiera('pool_create', true)
+  $mount_under_scratch_dirs       = hiera_array('mount_under_scratch_dirs', ["/tmp", "/var/tmp"])
```
Comment (kreczko) [inline manifests/params.pp]: `WARNING: double quoted string containing no variables on line 75`
should be
```puppet
$mount_under_scratch_dirs       = hiera_array('mount_under_scratch_dirs', ['/tmp', '/var/tmp'])
```

File: templates/50_singularity.config.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+SINGULARITY = <%= @singularity_path %>
+
+# Forces _all_ jobs to run inside singularity.
+SINGULARITY_JOB = <%= @force_singularity_jobs %>
+
+# Forces all jobs to use the given image.
+SINGULARITY_IMAGE_EXPR = <%= @singularity_image_expr %>
+
+# Paths to bind via user bind control. 
+SINGULARITY_BIND_EXPR = "<%= @singularity_bind_paths.flatten.join(", ") %>"
```
Comment (kreczko) [inline templates/50_singularity.config.erb]: Isn't the expression comma separated without space in between? I am not sure how HTCondor would behave in this situation.

Comment (olifre) [inline templates/50_singularity.config.erb]: You are right - interestingly, it works fine here, maybe path trimming happens at a later stage. I will address this problem (and the others) soon, thanks for checking carefully! 👍

**Comment by olifre** [comment]
> Could you maybe add an example set of parameters to the README? This would be especially useful for the bind paths.

Good idea, will do! 
Sadly, HTCondor singularity support is still in its very early stages, so I may even add the special bind paths which allow to run interactive jobs as an example. 
They are a bit ugly, since they hack around issues in HTCondor code - but might be helpful until HTCondor has fixed the problems.

**Comment by olifre** [comment]
All changes should be in, and I added a new section to the README in a separate commit.

**Comment by kreczko** [comment]
Thank you very much for the contribution!

**Pull request merged**